### PR TITLE
Big pack CLI tool

### PR DESCRIPTION
**Pull request by feliwir**
Adds a CLI tool to pack & unpack big archives. Similar to the linux **zip** command

### CODE REVIEW

File: src/OpenSage.FileFormats.Big/BigArchive.cs [hunk 1]
```diff
@@ -47,7 +47,11 @@ public BigArchive(string filePath, BigArchiveMode mode = BigArchiveMode.Update)
                 fileAccess,
                 fileShare));
 
-            Read();
+            // Read if the archive already exists
+            if(_stream.Length > 0 &&  mode != BigArchiveMode.Create)
```
Comment (tgjones) [inline src/OpenSage.FileFormats.Big/BigArchive.cs:51]: Why do we need the check for `_stream.Length > 0`?

File: src/OpenSage.FileFormats.Big/BigArchive.cs [hunk 2]
```diff
@@ -256,8 +260,10 @@ internal void WriteToDisk(bool forceWrite = false)
                     WriteFileTable(writer, dataStart);
                     WriteFileContent(writer);
                     _stream.Position = 0;
+                    var c = Encoding.ASCII.GetString(outArchive.ToArray());
```
Comment (tgjones) [inline src/OpenSage.FileFormats.Big/BigArchive.cs:263]: I presume this line was for debugging?

File: src/OpenSage.Tools.BigTool/Program.cs [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+
+        static void Main(string[] args)
+        {
+            var result = Parser.Default.ParseArguments<Options>(args)
+                .WithParsed(RunOptions);
+        }
+
+        static bool ValidateInputFiles(Options opts)
+        {
+            if(opts.List || opts.Update || opts.Delete )
```
Comment (tgjones) [inline src/OpenSage.Tools.BigTool/Program.cs:57]:
```suggestion
            if (opts.List || opts.Update || opts.Delete )
```

File: src/OpenSage.Tools.BigTool/Program.cs [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+            {
+                if(!File.Exists(opts.ArchiveName))
```
Comment (tgjones) [inline src/OpenSage.Tools.BigTool/Program.cs:59]:
```suggestion
                if (!File.Exists(opts.ArchiveName))
```

File: src/OpenSage.Tools.BigTool/Program.cs [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+                    }
+                }
+            }
+
+            return true;
+        }
+
+        static bool ValidateOptions(Options opts)
+        {
+            if(!opts.List && !opts.Extract && !opts.Files.Any())
```
Comment (tgjones) [inline src/OpenSage.Tools.BigTool/Program.cs:83]:
```suggestion
            if (!opts.List && !opts.Extract && !opts.Files.Any())
```

File: src/OpenSage.Tools.BigTool/Program.cs [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+            var archive = new BigArchive(opts.ArchiveName, BigArchiveMode.Update);
+            foreach (var entryName in opts.Files)
+            {
+                try
+                {
+                    var entry = archive.Entries.First(x => x.FullName == entryName);
+                    Console.WriteLine("deleting: {0}", entryName);
+                    entry.Delete();
+                }
+                catch(InvalidOperationException e)
```
Comment (tgjones) [inline src/OpenSage.Tools.BigTool/Program.cs:118]:
```suggestion
                catch (InvalidOperationException e)
```

File: src/OpenSage.Tools.BigTool/Program.cs [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+                }
+                AddDirectoryToArchive(archive, dir);
+            }
+        }
+
+        static void AddFileToArchive(BigArchive archive, string filepath, bool update = false)
+        {
+            BigArchiveEntry entry = null;
+
+            if(!update)
```
Comment (tgjones) [inline src/OpenSage.Tools.BigTool/Program.cs:141]:
```suggestion
            if (!update)
```

File: src/OpenSage.Tools.BigTool/Program.cs [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+
+        static void ExtractMode(Options opts)
+        {
+            var archive = new BigArchive(opts.ArchiveName, BigArchiveMode.Read);
+            foreach (var entry in archive.Entries)
+            {
+                using (var entryStream = entry.Open())
+                {
+                    var dirName = Path.GetDirectoryName(entry.FullName);
+                    if(!string.IsNullOrEmpty(dirName))
```
Comment (tgjones) [inline src/OpenSage.Tools.BigTool/Program.cs:213]:
```suggestion
                    if (!string.IsNullOrEmpty(dirName))
```

File: src/OpenSage.Tools.BigTool/Program.cs [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+                        Console.WriteLine("extracting: {0}", entry.FullName);
+                        entryStream.CopyTo(fileStream);
+                    }
+                }
+            }
+        }
+
+        static void RunOptions(Options opts)
+        {
+            if(!ValidateOptions(opts))
```
Comment (tgjones) [inline src/OpenSage.Tools.BigTool/Program.cs:229]:
```suggestion
            if (!ValidateOptions(opts))
```

File: src/OpenSage.Tools.BigTool/Program.cs [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+            {
+                return;
+            }
+
+            if (!ValidateInputFiles(opts))
+            {
+                return;
+            }
+
+            if(opts.List)
```
Comment (tgjones) [inline src/OpenSage.Tools.BigTool/Program.cs:239]:
```suggestion
            if (opts.List)
```

File: src/OpenSage.Tools.BigTool/Program.cs [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+            {
+                ListMode(opts);
+                return;
+            }
+
+            if(opts.Delete)
```
Comment (tgjones) [inline src/OpenSage.Tools.BigTool/Program.cs:245]:
```suggestion
            if (opts.Delete)
```

File: src/OpenSage.Tools.BigTool/Program.cs [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+            {
+                DeleteMode(opts);
+                return;
+            }
+
+            if(opts.Update)
```
Comment (tgjones) [inline src/OpenSage.Tools.BigTool/Program.cs:251]:
```suggestion
            if (opts.Update)
```

File: src/OpenSage.Tools.BigTool/Program.cs [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+            {
+                UpdateMode(opts);
+                return;
+            }
+
+            if(opts.Extract)
```
Comment (tgjones) [inline src/OpenSage.Tools.BigTool/Program.cs:257]:
```suggestion
            if (opts.Extract)
```

File: src/OpenSage.FileFormats.Big/BigArchive.cs (hunk 1 above)
Comment (feliwir) [reply src/OpenSage.FileFormats.Big/BigArchive.cs:51]: We always read the big file header, but i should probably add that check somewhere else

**Review by tgjones — APPROVED** [review]
Once it passes CI, this looks good to me.

**Pull request merged**